src/algebra/algebra/non_unital.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.algebra.basic
+
+/-!
+# Non-unital, non-associative algebras
+
+This files defines non-unital, non-associative algebras together with their morphisms and basic
+lemmas to make them usable.
+
+## Unitality, associativity and the legacy `algebra` class.
+
+The `non_unital_non_assoc_algebra` defined below can serve as a replacement for our older `algebra`
+class (see `non_unital_non_assoc_algebra.to_algebra`). The older algebra class is essentially just
+a more restrictive version of `non_unital_non_assoc_algebra` that can only be built on top of a
+unital, associative ring.
+
+The plan is to allow `non_unital_non_assoc_algebra` and `algebra` to coexist only pending the
+refactoring work to replace the definition of `algebra` with that of `non_unital_non_assoc_algebra`.
+
+## Main definitions
+
+  * `non_unital_non_assoc_algebra`
+  * `non_unital_non_assoc_algebra_hom`
+
+## Tags
+
+non-unital, non-associative, algebra
+-/
+
+universes u₁ u₂ u₃ u₄
+
+variables (R : Type u₁) (A : Type u₂) (B : Type u₃) (C : Type u₄)
+
+/-- A not-necessarily-unital, not-necessarily-associative algebra. -/
+@[protect_proj]
+class non_unital_non_assoc_algebra [semiring R] [non_unital_non_assoc_semiring A]
+  extends semimodule R A :=
+(smul_mul_assoc' : ∀ (t : R) (a b : A), (t • a) * b = t • (a * b))
+(mul_smul_comm' : ∀ (t : R) (a b : A), a * (t • b) = t • (a * b))
```

Comment:
I think this made it into a zulip message, but is worth recording in this PR too: these two conditions are respectively `[is_scalar_tower R A A]` and (modulo symmetry) `[smul_comm_class R A A]`.

Reply:
At least, it would be if we added `has_mul.to_has_scalar`, since `monoid.to_mul_action` can't apply here

Reply:
Yep, hadn't forgotten but good to get this reminder.

Reply:
I just might take advantage of this approach so I'll record the following snippet for emphasis:
```lean
import linear_algebra.basic

variables (R A : Type*) [semiring R] [non_unital_non_assoc_semiring A] [module R A]
variables (t : R) (a b : A)

example [smul_comm_class R A A] : a * (t • b) = t • (a * b) := (smul_comm t a b).symm

example [is_scalar_tower R A A] : (t • a) * b = t • (a * b) := smul_assoc t a b
```

Incidentally, the scalar action of `A` on itself is obtained from the combination of:

- `non_unital_non_assoc_semiring.to_mul_zero_class`
- `mul_zero_class.to_smul_with_zero`
- `smul_with_zero.to_has_scalar`

Reply:
I forgot that `smul_with_zero.to_has_scalar` applied here!

Reply:
Exactly right! Hence my remark at the top of #7847 about it being a plausible replacement for this class.

Reply:
(although we do face the problem of "exponential term size" if we go too far this way)